_P172_PZEM-004Tv3.ino

```diff
@@ -0,0 +1,873 @@
+            {
+                String json_topic = ControllerSettings.Publish;
+                parseControllerVariables(json_topic, event, false);
+                json_topic.replace("/%valname%", "");
+            
+                String values_in_json = "{";
+                values_in_json += to_json_object_value(ExtraTaskSettings.TaskDeviceValueNames[0],
+                                                       String(P172_PZEM004T_values_vector[currentIndex].voltage)) +
+                                  ",";
+                values_in_json += to_json_object_value(ExtraTaskSettings.TaskDeviceValueNames[1],
+                                                       String(P172_PZEM004T_values_vector[currentIndex].current)) +
+                                  ",";
+                values_in_json += to_json_object_value(ExtraTaskSettings.TaskDeviceValueNames[2],
+                                                       String(P172_PZEM004T_values_vector[currentIndex].power)) +
+                                  ",";
+                if (P172_PZEM004T_values_vector[currentIndex].energy >= 0)
+                {
+                    values_in_json += to_json_object_value(ExtraTaskSettings.TaskDeviceValueNames[3],
+                                                           String(P172_PZEM004T_values_vector[currentIndex].energy)) +
+                                  ",";
+                }
+                values_in_json += to_json_object_value(ExtraTaskSettings.TaskDeviceValueNames[4],
+                                                       String(P172_PZEM004T_values_vector[currentIndex].frequency)) +
+                                  ",";
+                values_in_json += to_json_object_value(ExtraTaskSettings.TaskDeviceValueNames[5],
+                                                   String(P172_PZEM004T_values_vector[currentIndex].pf)) +
+                              "}";
+                addLog(LOG_LEVEL_DEBUG, values_in_json);
+
+                MQTTpublish(event->ControllerIndex, json_topic.c_str(), values_in_json.c_str(), Settings.MQTTRetainFlag);
+            }
+            else
+            {
+                addLog(LOG_LEVEL_DEBUG_MORE, F("PZEM004Tv3: checkHostReachable = false"));
+            }
+            
+        }
+
+        UserVar[event->BaseVarIndex]     = P172_PZEM004T_values_vector[currentIndex].voltage;
+        UserVar[event->BaseVarIndex + 1] = P172_PZEM004T_values_vector[currentIndex].current;
+        UserVar[event->BaseVarIndex + 2] = P172_PZEM004T_values_vector[currentIndex].power;
+        if (P172_PZEM004T_values_vector[currentIndex].energy>=0)  
+        {
+            UserVar[event->BaseVarIndex + 3] = P172_PZEM004T_values_vector[currentIndex].energy;
+        }
+        UserVar[event->BaseVarIndex + 4] = P172_PZEM004T_values_vector[currentIndex].frequency;
+        UserVar[event->BaseVarIndex + 5] = P172_PZEM004T_values_vector[currentIndex].pf;

```

Comment:
Make the output values selectable and please don't use > 4 variables per plugin.
It really will make a lot of other things go wrong.
Please have a look at the Eastron, AcuEnergy and/or GPS plugin for some ideas on how to handle it.